src/smallOrLargeNodes/content.md

```diff
@@ -0,0 +1,293 @@
+
+Chances are that only some of your apps are affected, and potentially only a small number of replicas so that the apps as a whole stay up.
+
+Furthermore, there are most likely enough spare resources on the remaining nodes to accommodate the workload of the failed node, so that Kubernetes can reschedule all the pods and your apps return to a fully functional state relatively quickly.
+
+#### Allows high replication
+
+If you have replicated high-availability apps, and enough available nodes, the Kubernetes scheduler can assign each replica to a different node.
+
+> You can influence scheduler's the placement of pods with [node and pod affinities/anti-affinities](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) and with [taints and tolerations](https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/).
+
+This means that if a node fails, there is at most one replica affected and your app stays available.
+
+_Having seen the pros of using many small nodes, what are the cons?_
+
+### Cons
+
+#### Large number of nodes
+
+If you use smaller nodes, you naturally need more of them to achieve a given cluster capacity.
+
+_But large numbers of nodes can be a challenge for the Kubernetes control plane._
+
+For example, every node needs to be able to communicate with every other node, which makes the number of possible communication paths grow exponentially with the number of nodes — all of which has to be managed by the control plane.
+
+The node controller in the Kubernetes controller manager regularly has to iterate through all the nodes in the cluster and run health checks — more nodes mean thus more load for the node controller.
+
+More nodes mean also more load on the etcd database — this is because the kubelet and kube-proxy of every node register themselves as etcd clients and etcd must provide them with all updates to certain object types.
+
+In general, each worker node imposes some overhead on the system components on the master nodes.
+
+_Thus, the higher the number of worker nodes, the higher the performance requirements for the master nodes._
+
+This can be nicely seen in the [master node sizes](https://kubernetes.io/docs/setup/best-practices/cluster-large/#size-of-master-and-master-components) used by `kube-up` on cloud infrastructure:
+
+- [Google Cloud Platform](https://cloud.google.com/compute/docs/machine-types)
+  - 5 worker nodes → `n1-standard-1` master nodes
+  - 500 worker nodes → `n1-standard-32` master nodes
+- [Amazon Web Services](https://aws.amazon.com/ec2/instance-types/)
+  - 5 worker nodes → `m3.medium` master nodes
+  - 500 worker nodes → `c4.8xlarge` master nodes
+
+As you can see, for 500 worker nodes, the used master nodes have 32 and 36 CPU cores and 120 GB and 60 GB of memory, respectively.
+
+_These are pretty large machines!_
+
+Officially, Kubernetes claims to support clusters with [up to 5000 nodes](https://kubernetes.io/docs/setup/best-practices/cluster-large/).
```

Comment:
Fun fact, a few companies in China are really pushing the boundaries for k8s clusters.

[Netease: 30,000 nodes for a single cluster](https://kubernetes.io/case-studies/netease/)
> NetEase internal cloud platform—which also leverages the CNCF projects Prometheus, Envoy, Harbor, gRPC, and Helm—runs 10,000 nodes in a production cluster and can support up to 30,000 nodes in a cluster.

Alibaba is using the virtual kubelet + their equivalent fargate service to do immediate scaling for the k8s cluster with tens of thousands of nodes.
More:
- https://medium.com/@Alibaba_Cloud/from-containers-to-cloud-native-ultimate-elasticity-a649f033bf6
- https://www.youtube.com/watch?v=v9cwYvuzROs

Reply:
Nice, that's interesting